tests/cpp/test_remove_bcast_squeeze.cpp

```diff
@@ -0,0 +1,281 @@
+ */
+// clang-format on
+#include <csrc/exceptions.h>
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+
+#include <ir/all_nodes.h>
+#include <ir/utils.h>
+#include <ops/all_ops.h>
+#include <preseg_passes/optimization_pass.h>
+#include <preseg_passes/pre_segmenter.h>
+#include <tests/cpp/utils.h>
+#include <tests/cpp/validator.h>
+
+#include <torch/torch.h>
+
+namespace nvfuser {
+
+using RemoveBcastSqueezeTest = NVFuserTest;
+
+namespace {
+std::pair<bool, bool> hasBcastSqueeze(Fusion* fusion) {
+  bool has_bcast = false;
+  bool has_squeeze = false;
+  for (auto expr : fusion->exprs()) {
+    if (dynamic_cast<BroadcastOp*>(expr)) {
+      has_bcast = true;
+    }
+    if (dynamic_cast<SqueezeOp*>(expr)) {
+      has_squeeze = true;
+    }
+  }
+  return std::make_pair(has_bcast, has_squeeze);
+}
+
+inline TensorView* makeBroadcastTensor(
+    const std::vector<bool>& is_broadcast_dim,
+    DataType input_dtype = DataType::Float) {
+  std::vector<IterDomain*> out_domain;
+  out_domain.reserve(is_broadcast_dim.size());
+  for (auto is_broadcast : is_broadcast_dim) {
+    out_domain.push_back(
+        IterDomainBuilder(
+            FusionGuard::getCurFusion()->zeroVal(),
+            is_broadcast ? FusionGuard::getCurFusion()->oneVal()
+                         : IrBuilder::create<Val>(DataType::Index))
+            .iter_type(is_broadcast ? IterType::Broadcast : IterType::Iteration)
+            .build());
+  }
+  return IrBuilder::create<TensorView>(
+      IrBuilder::create<TensorDomain>(
+          out_domain, TensorDomain::getContiguityFilledWith(out_domain, true)),
+      input_dtype);
+}
+
+} // namespace
+
+TEST_F(RemoveBcastSqueezeTest, BcastSqueeze) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+  DataType input_dtype = DataType::Float;
+  auto tv0 = makeContigTensor(2, input_dtype);
+  fusion->addInput(tv0);
+  auto tv1 = set(tv0);
+  auto tv2 = broadcast(tv1, {false, false, true});
+  auto tv3 = squeeze(tv2, std::vector<bool>{false, false, true});
+  auto tv4 = set(tv3);
+  fusion->addOutput(tv4);
+
+  // preseg_passes should remove both broadcast and squeeze
+  preseg_passes::OptimizationPass<preseg_passes::PreSegmenter>::runPass(
+      fusion.get());
+  auto [has_bcast, has_squeeze] = hasBcastSqueeze(fusion.get());
+  ASSERT_FALSE(has_bcast);
+  ASSERT_FALSE(has_squeeze);
+}
+
+TEST_F(RemoveBcastSqueezeTest, BcastSqueezeUnmatchedDim) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+  DataType input_dtype = DataType::Float;
+  auto tv0 = makeContigTensor(2, input_dtype);
+  fusion->addInput(tv0);
+  auto tv1 = set(tv0);
+  auto tv2 = broadcast(tv1, {false, false, true, true});
+  auto tv3 = squeeze(tv2, std::vector<bool>{false, false, false, true});
+  auto tv4 = set(tv3);
+  fusion->addOutput(tv4);
+
+  // preseg_passes shouldn't remove either broadcast or squeeze
+  // becuase broadcast dim doesn't match with squeeze dim
```

Comment:
Should we consider still doing a replacement in cases like this where the flag patterns overlap but are not equal? In this case we would replace
```c++
auto tv2 = broadcast(tv1, {false, false, true, true});
auto tv3 = squeeze(tv2, std::vector<bool>{false, false, false, true});
```
with
```c++
auto tv2 = broadcast(tv1, {false, false, true, false});
```
and that would still remove a bcast+squeeze on the inner dim.

Reply:
after this there is still a broadcast dim, may not bring much benefit. so it's not removed in this PR but we can reconsider later.